File: 102-Ecological Bin Packing/band.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;


void BGC(vector<int>*, vector<int>*, vector<int>*, int*, string*);
void BCG(vector<int>*, vector<int>*, vector<int>*, int*, string*);


int main()
{
    vector<int> bin_one, bin_two, bin_three;
    int int_one, int_two, int_three, int_four, int_five, int_six, int_seven, int_eight, int_nine, minSum = 0;
    string order;

    while(cin>>int_one>>int_two>>int_three>>int_four>>int_five>>int_six>>int_seven>>int_eight>>int_nine)
    {
        bin_one.clear();    bin_two.clear();    bin_three.clear();

        bin_one.push_back(int_one);     bin_one.push_back(int_two);     bin_one.push_back(int_three);
        bin_two.push_back(int_four);    bin_two.push_back(int_five);    bin_two.push_back(int_six);
        bin_three.push_back(int_seven);   bin_three.push_back(int_eight); bin_three.push_back(int_nine);

        int minSum = -1;    string order;
        BGC(&bin_one, &bin_two, &bin_three, &minSum, &order);
        BCG(&bin_one, &bin_two, &bin_three, &minSum, &order);
        cout<<order<<" "<<minSum<<endl;
    }

    return 0;
}


void BGC(vector<int>* bin_one, vector<int>* bin_two, vector<int>* bin_three, int* minSum, string* order)
{
    int sum = 0;

    for(int a=0; a<3; a++)
    {
        if(a == 0)
        {
            sum = bin_one[0][1]+bin_one[0][2]+bin_two[0][0]+bin_two[0][2]+bin_three[0][0]+bin_three[0][1];
            *minSum = sum;  *order = "BGC";
        }
        else if(a == 1)
        {
            sum = bin_one[0][0]+bin_one[0][2]+bin_two[0][0]+bin_two[0][1]+bin_three[0][1]+bin_three[0][2];
            if(*minSum != sum)
            {
                *minSum = min(*minSum, sum);
                if(*minSum == sum)
                {
                    *order = "GCB";
                }
            }
            else
            {
                string temp = "GCB";    *order = min(temp, *order);
            }   
        }
        else if(a == 2)
        {
            sum = bin_one[0][0]+bin_one[0][1]+bin_two[0][1]+bin_two[0][2]+bin_three[0][0]+bin_three[0][2];
            if(*minSum != sum)
            {
                *minSum = min(*minSum, sum);
                if(*minSum == sum)
                {
                    *order = "CBG";
                }
            }
            else
            {
                string temp = "CBG";        *order = min(temp, *order);
            }
        }
    }
    
    return;
}


void BCG(vector<int>* bin_one, vector<int>* bin_two, vector<int>* bin_three, int* minSum, string* order)
{
    int sum = 0;

    for(int a=0; a<3; a++)
    {
        if(a == 0)
        {
            sum = bin_one[0][1]+bin_one[0][2]+bin_two[0][0]+bin_two[0][1]+bin_three[0][0]+bin_three[0][2];
            if(*minSum != sum)
            {
                *minSum = min(sum, *minSum);
                if(*minSum == sum)
                {
                    *order = "BCG";
                }
            }
            else
            {
                string temp = "BCG";        *order = min(temp, *order);   
            }
        }
        else if(a == 1)
        {
            sum = bin_one[0][0]+bin_one[0][1]+bin_two[0][0]+bin_two[0][2]+bin_three[0][1]+bin_three[0][2];
            if(*minSum != sum)
            {
                *minSum = min(sum, *minSum);
                if(*minSum == sum)
                {
                    *order = "CGB";
                }
            }
            else
            {
                string temp = "CGB";    *order = min(temp, *order);
            }
            
        }
        else if(a == 2)
        {
            sum = bin_one[0][0]+bin_one[0][2]+bin_two[0][1]+bin_two[0][2]+bin_three[0][0]+bin_three[0][1];
            if(*minSum != sum)
            {
                *minSum = min(*minSum, sum);
                if(*minSum == sum)
                {
                    *order = "GBC";
                }
            }
            else
            {
                string temp = "GBC";        *order = min(temp, *order);
            }
        }
    }
    
    return;
}

File: 102-Ecological Bin Packing/band.py
import sys


def BGC(bin_one, bin_two, bin_three):
    sum = 0
    maxSum = -1
    order = ''

    for a in range(3):
        if a == 0:
            sum = bin_one[1]+bin_one[2]+bin_two[0]+bin_two[2]+bin_three[0]+bin_three[1]
            maxSum = sum
            order = "BGC"
        elif a == 1:
            sum = bin_one[0]+bin_one[2]+bin_two[0]+bin_two[1]+bin_three[1]+bin_three[2]
            if maxSum != sum:
                maxSum = min(maxSum, sum)
                if maxSum == sum:
                    order = "GCB"
            else:
                order = min("GCB", order)
        elif a == 2:
            sum = bin_one[0]+bin_one[1]+bin_two[1]+bin_two[2]+bin_three[0]+bin_three[2]
            if maxSum != sum:
                maxSum = min(maxSum, sum)
                if maxSum == sum:
                    order = "CBG"
            else:
                order = min("CBG", order)

    return maxSum, order


def BCG(bin_one, bin_two, bin_three, maxSum, order):
    sum = 0

    for a in range(3):
        if a == 0:
            sum = bin_one[1]+bin_one[2]+bin_two[0]+bin_two[1]+bin_three[0]+bin_three[2]
            if maxSum != sum:
                maxSum = min(maxSum, sum)
                if maxSum == sum:
                    order = "BCG"
            else:
                order = min(order, "BCG")
        elif a == 1:
            sum = bin_one[0]+bin_one[1]+bin_two[0]+bin_two[2]+bin_three[1]+bin_three[2]
            if maxSum != sum:
                maxSum = min(maxSum, sum)
                if maxSum == sum:
                    order = "CGB"
            else:
                order = min("CGB", order)
        elif a == 2:
            sum = bin_one[0]+bin_one[2]+bin_two[1]+bin_two[2]+bin_three[0]+bin_three[1]
            if maxSum != sum:
                maxSum = min(maxSum, sum)
                if maxSum == sum:
                    order = "GBC"
            else:
                order = min("GBC", order)

    print(order, maxSum)

    return


if __name__ == "__main__":
    for a in sys.stdin:
        line = a.split()

        bin_one = []
        [bin_one.append(int(line[b])) for b in range(0, 3)]

        bin_two = []
        [bin_two.append(int(line[b])) for b in range(3, 6)]

        bin_three = []
        [bin_three.append(int(line[b])) for b in range(6, 9)]

        maxSum, order = BGC(bin_one, bin_two, bin_three)
        BCG(bin_one, bin_two, bin_three, maxSum, order)

    exit(0)
